File: 7-2.py
#!/usr/bin/env python

crabs = [int(crab) for crab in open('7.input').readlines()[0].split(",")]

leftmost = min(crabs)
rightmost = max(crabs)

cost = [0] * (rightmost + 1)

for gather in range(leftmost, rightmost + 1):
    cost[gather] = int(sum([abs(gather - crab) / 2 * (1 + abs(gather - crab)) for crab in crabs]))

print(min(cost))

File: 7-2.rb
#!/usr/bin/env ruby

crabs = File.read('7.input').strip.split(",").map(&:to_i)

leftmost = crabs.min
rightmost = crabs.max

cost = [0] * (rightmost + 1)

leftmost.upto(rightmost) do |gather|
  cost[gather] = (crabs.map do |crab|
                    n = (gather - crab).abs
                    (n / 2.0 * (1 + n)).to_i
                  end).sum
end

print cost.min, "\n"
